pkg/utils/lifecycle/lifecycle_test.go

```diff
@@ -0,0 +1,321 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package lifecycle
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"sigs.k8s.io/kwok/pkg/apis/internalversion"
+)
+
+func TestListAllPossibleStages(t *testing.T) {
+	stages := []*internalversion.Stage{
+		{
+			Spec: internalversion.StageSpec{
+				Selector: &internalversion.StageSelector{
+					MatchLabels: map[string]string{
+						"a": "b",
+						"c": "d",
+					},
+				},
+			},
+		},
+		{
+			Spec: internalversion.StageSpec{
+				Selector: &internalversion.StageSelector{
+					MatchAnnotations: map[string]string{
+						"e": "f",
+						"g": "h",
+					},
+				},
+			},
+		},
+	}
+	lc, err := NewLifecycle(stages)
+	if err != nil {
+		t.Fatal("Could not create a new lifecycle:", err)
+	}
+	label := labels.Set{
+		"a": "b",
+		"c": "d",
+	}
+	annotation := labels.Set{
+		"e": "f",
+		"g": "h",
+	}
+	var data interface{}
+	var possibleStages []*Stage
+	var matches bool
+	possibleStages, err = lc.ListAllPossible(label, annotation, data)
+	if err != nil {
+		t.Fatal("Could not list all possible Stages:", err)
+	}
+	for i := 0; i < len(possibleStages); i++ {
+		matches, err = possibleStages[i].match(label, annotation, data)
+		if err != nil {
+			t.Fatal("Could not match stages")
+		}
+		if !matches {
+			t.Fatal("Expected the stage to be matched")
+		}
```

Comment:
This is not a good way to check, it will never find a problem, we should compare it to a fixed value.

Reply:
@wzshiming do you mean that we should compare `possibleStages[i].matchLabels` with `label` and `possibleStages[i].matchAnnotations` with `annotation` ?

Reply:
In other words, even if this function returns an empty data exception, this case will not detect any problems,

Reply:
@wzshiming I have made the changes and now we also check for empty list of stages